### PR TITLE
Update the default behavior of `zero_grad(set_to_none=None)` to align with PyTorch

### PR DESCRIPTION
**Pull request by yongchanghao**
Now, the behavior of the wrapped optimizer is that the gradient is cleared by default when `set_to_none=None`. This aligns with `torch.optim.Optimizer` and saves memory.

# What does this PR do?

Fixes https://github.com/huggingface/accelerate/issues/2471

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by yongchanghao** [comment]
This is a proposed update for issue #2471

**Review by muellerzr — APPROVED** [review]
Thanks! Since we align with PyTorch and was over a year ago I think it's safe to just pass this in, however I'll make some tweaks to the title etc in case users dig for this

**Review by BenjaminBossan — APPROVED** [review]
LGTM too. We should mention this switch in the next release notes.

**Comment by muellerzr** [comment]
@yongchanghao can you rebase from main to check if these failing tests pass? Thanks!

**Comment by yongchanghao** [comment]
I have sync the code with the main branch.

**Comment by muellerzr** [comment]
Thanks!
